thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/wrapper/AnomalyDetectorWrapper.java

```diff
@@ -290,6 +291,16 @@ long getLastTimeStamp() {
           DateTime startTime = endTime.minus(windowSizePeriod);
           monitoringWindows.add(new Interval(startTime, endTime));
         }
+        if (isSpeedUp) {
```

Comment:
Why we generate the window larger than endTime?
Should we fix the window generation logic instead of patching here?

Reply:
Good point. Changed it.

Reply:
Thanks for the change. I mean can we change getMonitoringWindows()?
This is the original place we generate the windows.

Reply:
Yes, that's where I added the check.

Reply:
Sorry I didn't notice that.
Since here is the place all the windows are generated. I don't see a reason why you need to modify existing windows instead of adding the check when generating the windows.

Can we just change L289 from
          DateTime endTime = monitoringEndTime.minus(windowDelayPeriod);
To:
        DateTime windowEndTime = Math.min(monitoringEndTime.minus(windowDelayPeriod), endTime)?

Reply:
We cannot change the endTime here directly, as it can generate windows that overlaps the previous windows. Either we change the endTime of latest interval or we need to change the logic of generating the windows completely.

Reply:
Why it can generate windows that overlaps the previous windows? Can we sync offline?

Reply:
I got your point. Then we can do a little trick - modify endTime after using it.
How about the following? @vincentchenjl 

`List<Interval> monitoringWindows = new ArrayList<>();
List<DateTime> monitoringWindowEndTimes = getMonitoringWindowEndTimes();
DateTime detectionEndTime = new DateTime(endTime, dateTimeZone).minus(windowDelayPeriod);
for (DateTime monitoringEndTime : monitoringWindowEndTimes) {
  DateTime endTime = monitoringEndTime.minus(windowDelayPeriod);
  DateTime startTime = endTime.minus(windowSizePeriod);
  endTime = endTime.isAfter(detectionEndTime) ? detectionEndTime : endTime;
  monitoringWindows.add(new Interval(startTime, endTime));
}
`

Reply:
Done.

Reply:
Great.  Let's not fix for the symptom. Fix for the root cause, as early as possible.